File: client/app/src/desktopMain/kotlin/com/jahirtrap/cconnect/terminal/OsIcons.kt
package com.jahirtrap.cconnect.terminal

import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import com.composables.icons.fontawesome.FontAwesome
import com.composables.icons.fontawesome.brands.Android
import com.composables.icons.fontawesome.brands.Apple
import com.composables.icons.fontawesome.brands.Centos
import com.composables.icons.fontawesome.brands.Fedora
import com.composables.icons.fontawesome.brands.Linux
import com.composables.icons.fontawesome.brands.RaspberryPi
import com.composables.icons.fontawesome.brands.Redhat
import com.composables.icons.fontawesome.brands.Suse
import com.composables.icons.fontawesome.brands.Ubuntu
import com.composables.icons.fontawesome.brands.Windows
import com.composables.icons.lucide.Lucide
import com.composables.icons.lucide.SquareTerminal

private fun normalize(os: String?) = os?.trim()?.lowercase().orEmpty()

fun iconForOs(os: String?): ImageVector = when (normalize(os)) {
    "" -> Lucide.SquareTerminal
    "windows" -> FontAwesome.Brands.Windows
    "macos", "darwin", "apple", "ios" -> FontAwesome.Brands.Apple
    "android" -> FontAwesome.Brands.Android
    "ubuntu" -> FontAwesome.Brands.Ubuntu
    "fedora" -> FontAwesome.Brands.Fedora
    "centos" -> FontAwesome.Brands.Centos
    "rhel", "redhat", "red-hat", "rocky", "almalinux", "alma" -> FontAwesome.Brands.Redhat
    "opensuse", "opensuse-leap", "opensuse-tumbleweed", "suse", "sles" -> FontAwesome.Brands.Suse
    "raspbian", "raspberrypi", "raspberry-pi" -> FontAwesome.Brands.RaspberryPi
    else -> FontAwesome.Brands.Linux
}

fun colorForOs(os: String?): Color? = when (normalize(os)) {
    "windows" -> Color(0xFF00A4EF)
    "macos", "darwin", "apple", "ios" -> Color(0xFFA2AAAD)
    "android" -> Color(0xFF3DDC84)
    "ubuntu" -> Color(0xFFE95420)
    "debian" -> Color(0xFFA81D33)
    "fedora" -> Color(0xFF51A2DA)
    "centos" -> Color(0xFF932279)
    "rhel", "redhat", "red-hat" -> Color(0xFFEE0000)
    "rocky" -> Color(0xFF10B981)
    "almalinux", "alma" -> Color(0xFF0FB37D)
    "opensuse", "opensuse-leap", "opensuse-tumbleweed", "suse", "sles" -> Color(0xFF73BA25)
    "arch", "archlinux" -> Color(0xFF1793D1)
    "manjaro" -> Color(0xFF35BF5C)
    "endeavouros" -> Color(0xFF7F3FBF)
    "linuxmint", "mint" -> Color(0xFF87CF3E)
    "pop", "popos", "pop_os" -> Color(0xFF48B9C7)
    "kali", "kalilinux", "kali-linux" -> Color(0xFF367BF0)
    "alpine" -> Color(0xFF0D597F)
    "raspbian", "raspberrypi", "raspberry-pi" -> Color(0xFFC51A4A)
    "freebsd" -> Color(0xFFAB2B28)
    "openbsd" -> Color(0xFFF2CA30)
    "nixos", "nix" -> Color(0xFF5277C3)
    "gentoo" -> Color(0xFF54487A)
    "elementary" -> Color(0xFF64BAFF)
    "void", "voidlinux" -> Color(0xFF478061)
    "linux" -> Color(0xFFFCC624)
    else -> null
}

File: client/src/lib/design/osIcons.ts
import Almalinux from "~icons/devicon/almalinux";
import Android from "~icons/logos/android-icon";
import Apple from "~icons/logos/apple";
import Archlinux from "~icons/logos/archlinux";
import Artixlinux from "~icons/devicon/artixlinux";
import Centos from "~icons/devicon/centos";
import Debian from "~icons/logos/debian";
import Elementary from "~icons/logos/elementary";
import Fedora from "~icons/logos/fedora";
import Freebsd from "~icons/logos/freebsd";
import Gentoo from "~icons/devicon/gentoo";
import Kalilinux from "~icons/devicon/kalilinux";
import Linux from "~icons/logos/linux-tux";
import Manjaro from "~icons/logos/manjaro";
import Mint from "~icons/logos/linux-mint";
import Nixos from "~icons/devicon/nixos";
import Opensuse from "~icons/devicon/opensuse";
import Raspberrypi from "~icons/logos/raspberry-pi";
import Redhat from "~icons/devicon/redhat";
import Rocky from "~icons/devicon/rockylinux";
import Ubuntu from "~icons/logos/ubuntu";
import Void from "~icons/logos/void";
import Windows10 from "~icons/devicon/windows8";
import Windows11 from "~icons/devicon/windows11";
import type { Component } from "svelte";

const normalize = (os: string | null | undefined) => os?.trim().toLowerCase() ?? "";

const BRANDS: Record<string, Component> = {
  windows: Windows11,
  "windows-11": Windows11,
  "windows-10": Windows10,
  macos: Apple,
  darwin: Apple,
  apple: Apple,
  ios: Apple,
  android: Android,
  ubuntu: Ubuntu,
  debian: Debian,
  fedora: Fedora,
  centos: Centos,
  rhel: Redhat,
  redhat: Redhat,
  "red-hat": Redhat,
  rocky: Rocky,
  rockylinux: Rocky,
  almalinux: Almalinux,
  alma: Almalinux,
  opensuse: Opensuse,
  "opensuse-leap": Opensuse,
  "opensuse-tumbleweed": Opensuse,
  suse: Opensuse,
  sles: Opensuse,
  arch: Archlinux,
  archlinux: Archlinux,
  endeavouros: Archlinux,
  artix: Artixlinux,
  artixlinux: Artixlinux,
  manjaro: Manjaro,
  linuxmint: Mint,
  mint: Mint,
  elementary: Elementary,
  void: Void,
  voidlinux: Void,
  freebsd: Freebsd,
  kali: Kalilinux,
  kalilinux: Kalilinux,
  "kali-linux": Kalilinux,
  nixos: Nixos,
  nix: Nixos,
  gentoo: Gentoo,
  raspbian: Raspberrypi,
  raspberrypi: Raspberrypi,
  "raspberry-pi": Raspberrypi,
};

export const osIcon = (os: string | null | undefined): Component | null => {
  const key = normalize(os);
  if (!key) return null;
  return BRANDS[key] ?? Linux;
};
